File: Stacks-And-Queues/Challenge-5.js
// function sortStack(mainStack) {
//   const tempStack = [];

//   while (mainStack.length > 0) {
//     const temp = mainStack.pop();
//     while (tempStack.length > 0 && tempStack[tempStack.length - 1] > temp) {
//       mainStack.push(tempStack.pop());
//     }
//     tempStack.push(temp);
//   }

//   while (tempStack.length > 0) {
//     mainStack.push(tempStack.pop());
//   }
// }

// const unsortedStack = [4, 3, 2, 5, 1];
// sortStack(unsortedStack);
// console.log(unsortedStack);

class SortStack {
  constructor() {
    this.mainStack = [];
    this.tempStack = [];
  }
  sort() {
    while (this.mainStack.length > 0) {
      const temp = this.mainStack.pop();

      while (
        this.tempStack.length > 0 &&
        this.tempStack[this.tempStack.length - 1] < temp
      ) {
        this.mainStack.push(this.tempStack.pop());
      }
      this.tempStack.push(temp);
    }
    while (this.tempStack.length > 0) {
      this.mainStack.push(this.tempStack.pop());
    }
  }
}

const sortStack = new SortStack();
sortStack.mainStack.push(3, 4, 5, 1, 2);
sortStack.sort();
console.log(sortStack.mainStack);

File: Stacks-And-Queues/Challenge-5.py
# def sort_stack(stack):
#   temp_stack = []

#   while stack:
#     temp = stack.pop()

#     while temp_stack and temp_stack[-1] < temp:
#       stack.append(temp_stack.pop())

#     temp_stack.append(temp)

#   while temp_stack:
#     stack.append(temp_stack.pop())

# stack = [3,4,1,5,2]
# sort_stack(stack)
# print(stack)

class SortedStack:
  def __init__(self):
    self.stack = []
    self.temp_stack = []

  def push(self,value):

    while self.stack and self.stack[-1] > value:
      self.temp_stack.append(self.stack.pop())

    self.stack.append(value)

    while self.temp_stack:
      self.stack.append(self.temp_stack.pop())

  def pop(self):
    if not self.is_empty():
      return self.stack.pop()
    else:
      print("Stack is empty.")
      return None

  def peek(self):
    if not self.is_empty():
      return self.stack[-1]
    else:
      print('Stack is empty.')
      return None

  def is_empty(self):
    return not bool(self.stack)

  def display(self):
    print(self.stack)

sorted_stack = SortedStack()
sorted_stack.push(3)
sorted_stack.push(1)
sorted_stack.push(4)
sorted_stack.push(2)
sorted_stack.push(5)
sorted_stack.display()
